protocol-units/settlement/mcr/client/src/stub.rs

```diff
@@ -0,0 +1,149 @@
+use crate::{CommitmentStream, McrSettlementClientOperations};
+use movement_types::BlockCommitment;
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::sync::{mpsc, RwLock};
+
+pub struct McrSettlementClient {
+	commitments: Arc<RwLock<HashMap<u64, BlockCommitment>>>,
+	stream_sender: mpsc::Sender<Result<BlockCommitment, anyhow::Error>>,
+	// todo: this is logically dangerous, but it's just a stub
+	stream_receiver: Arc<RwLock<mpsc::Receiver<Result<BlockCommitment, anyhow::Error>>>>,
+    pub current_height: Arc<RwLock<u64>>,
+    pub block_lead_tolerance: u64
+}
+
+impl McrSettlementClient {
+	pub fn new() -> Self {
+		let (stream_sender, receiver) = mpsc::channel(10);
+		McrSettlementClient {
+			commitments: Arc::new(RwLock::new(HashMap::new())),
+			stream_sender,
+			stream_receiver: Arc::new(RwLock::new(receiver)),
+            current_height: Arc::new(RwLock::new(0)),
+            block_lead_tolerance: 16,
+		}
+	}
+}
+
+
+#[async_trait::async_trait]
+impl McrSettlementClientOperations for McrSettlementClient {
+	async fn post_block_commitment(
+		&self,
+		block_commitment: BlockCommitment,
+	) -> Result<(), anyhow::Error> {
+
+        let height = block_commitment.height;
+
+        {
+            let mut commitments = self.commitments.write().await;
+            commitments.insert(block_commitment.height, block_commitment.clone());
+            self.stream_sender.send(Ok(block_commitment)).await?; // Simulate sending to the stream.
+        }
+
+        {
+            let mut current_height = self.current_height.write().await;
+            if height > *current_height {
+                *current_height = height;
+            }
+        }
+
+		Ok(())
+	}
+
+    async fn post_block_commitment_batch(&self, block_commitment: Vec<BlockCommitment>) -> Result<(), anyhow::Error> {
+        for commitment in block_commitment {
+            self.post_block_commitment(commitment).await?;
+        }
+        Ok(())
+    }
+
+	async fn stream_block_commitments(&self) -> Result<CommitmentStream, anyhow::Error> {
+		let receiver = self.stream_receiver.clone();
+		let stream = async_stream::try_stream! {
+			let mut receiver = receiver.write().await;
```

Comment:
The Mutex is write-locked for the whole duration of the stream. So if you call `stream_block_commitments` twice, the second stream will be locked and send nothing.
I think this pattern to return a stream in a trait that can be called several times is problematic because you'll need to implement sort of broadcast channel under and not all impl can do this easily.
To avoid that, I separate the post from the received part. The post call is implemented in the trait and the stub constructor return the stream so this way we are sure that the stream is created once, and it removes all Mutex need. In the end, we need only one stream.

Reply:
@musitdev I think we should have the stream type somehow available from the client trait (as an associated type?) to be able to mock the client in tests and have the same generic code for production.

Reply:
You have it from the constructor. One trait to post and one trait to stream. You get them differently.
For example:
```
impl McrSettlementClient {
	pub fn new() -> Result<(Self, CommitmentStream), anyhow::Error>  {
		...
	}
}
let (client, stream) = McrSettlementClient::new();
```

Reply:
A more idiomatic way can be 
```
impl McrSettlementClient {
	pub fn build() -> Result<(impl McrSettlementClientOperations, CommitmentStream), anyhow::Error>
	{
		todo!()
	}
}
```

Reply:
Then it will have to be passed in as [another parameter](https://github.com/movementlabsxyz/movement/blob/843822aa2cd6d9d9c88df9abc6c067bd64d07a6d/networks/suzuka/suzuka-full-node/src/partial.rs#L41) when constructing the full node. Also, what should happen if the stream handle is dropped (because, perhaps the consumer is only interested in submitting)?

Reply:
> I think this pattern to return a stream in a trait that can be called several times is problematic because you'll need to implement sort of broadcast channel under and not all impl can do this easily.

This is a problem with the stub (and can be solved by "don't do that, then" for now), but in a protocol like gRPC, it wouldn't be a problem to establish multiple subscriptions except duplicate traffic on the wire. Maybe we should redesign the API to not allow this accidentally.

Reply:
In fact, I was wrong. I focus on the implementation and not of the interface. What we should do is return the trait from the constructor. The stream creation can be differed. What the implementation should do it to take care if the creation is not possible (for example only one creation like in the stub) is to return an error.
So the constructor should be 
```
impl McrSettlementClient {
	pub fn build() -> impl McrSettlementClientOperation
	{
		//current code
	}
}
```
If McrSettlementClientOperation is send+sync it should be the same.

Reply:
In fact as stub impl is for test purpose you can do has you want. I was more thinking as the way I'm doing.